File: liskov/code.js
class Bird {
  fly() {
    return "I can fly!";
  }
}

class Penguin extends Bird {
  fly() {
    throw new Error("Penguins can't fly!");
  }
}

function makeBirdFly(bird) {
  console.log(bird.fly());
}

const sparrow = new Bird();
const penguin = new Penguin();

console.log(makeBirdFly(sparrow));
console.log(makeBirdFly(penguin));

class Bird {
  fly() {
    return "I can fly!";
  }
}

class Penguin extends Bird {
  // ❌ Violates LSP
  fly() {
    throw new Error("Penguins can't fly!");
  }
}

// Usage
function makeBirdFly(bird) {
  console.log(bird.fly());
}

const sparrow = new Bird();
const penguin = new Penguin();

console.log(makeBirdFly(sparrow)); // ✅ "I can fly!"
console.log(makeBirdFly(penguin)); // ❌ Error: Penguins can't fly!

class NewBird {
  makeSound() {
    return "Some generic bird sound";
  }
}

class FlyingBird extends NewBird {
  fly() {
    return "I can fly!";
  }
}

class NonFlyingBird extends NewBird {
  fly() {
    return "I can't fly!";
  }
}

class Sparrow extends FlyingBird {
  makeSound() {
    return "Chirp!";
  }
}

class Penguin extends NonFlyingBird {
  makeSound() {
    return "Honk!";
  }
}

function makeBirdFly(bird) {
  console.log(bird.fly());
}

const sparrow1 = new Sparrow();
const penguin1= new Penguin();

console.log(makeBirdFly(sparrow1)); 
console.log(makeBirdFly(penguin1)); 

File: design-principles/liskov/code.py
class Bird:
    def fly(self):
        return "I can fly!"

class Penguin(Bird):  
    def fly(self):
        raise NotImplementedError("Penguins can't fly!")


def make_bird_fly(bird: Bird):
    print(bird.fly())


sparrow = Bird()
penguin = Penguin()

print(make_bird_fly(sparrow))   
print(make_bird_fly(penguin))  

from abc import ABC, abstractmethod

class Bird(ABC):
    @abstractmethod
    def make_sound(self):
        pass

class FlyingBird(Bird):
    def fly(self):
        return "I can fly!"

class NonFlyingBird(Bird):
    def fly(self):
        return "I can't fly!"

class Sparrow(FlyingBird):
    def make_sound(self):
        return "Chirp!"

class Penguin(NonFlyingBird):
    def make_sound(self):
        return "Honk!"

def make_bird_fly(bird: FlyingBird):
    print(bird.fly())

sparrow = Sparrow()
penguin = Penguin()

print(make_bird_fly(sparrow))
